pairing/bn256/point.go

```diff
@@ -108,15 +108,16 @@ func (p *pointG1) Mul(s kyber.Scalar, q kyber.Point) kyber.Point {
 
 func (p *pointG1) MarshalBinary() ([]byte, error) {
 	n := p.ElementSize()
-	p.g.MakeAffine()
+	pgtemp := *p.g
```

Comment:
There needs to be a comment here explaining why we take a copy. I propose:
```
// Take a copy so that p is not written to, so calls to MarshalBinary
// are threadsafe.

Reply:
The real problem here is that there's an undocumented assumption that MarshalBinary is read only. This change brings bn256 back in line with the undocumented assumption, so that onet is no longer getting a data race.

It is not clear where/how to document the assumption, since MarshalBinary is an interface we inherit from Go itself. (And Go's definition is not even clear on this point: https://golang.org/pkg/encoding/#BinaryMarshaler )

Reply:
Yes, correct! We could put it here:

https://github.com/dedis/kyber/blob/master/encoding.go#L13

And add a comment in there stating that the implementers of `Marshaling` should make it thread-safe.

Bad joke: let's add `Promises` to go - I miss them from javascript. They make life so much easier.